File: jscomp/others/belt_internalSetInt.ml
# 5 "others/internal_set.cppo.ml"
type value = int
module S = Belt_SortArrayInt


# 12 "others/internal_set.cppo.ml"
module N = Belt_internalAVLset
module A = Belt_Array 


type t = value N.t

let rec has (t : t) (x : value)  =
  match t with 
  | None -> false
  | Some n  ->                
    let v = n.value in 
    x = v || has (if x < v then n.left else n.right) x


let rec compareAux e1 e2  =
    match e1,e2 with 
    | h1::t1, h2::t2 ->
        let (k1 : value) ,k2 = h1.N.value , h2.N.value  in 
        if k1 = k2 then  
          compareAux 
            (N.stackAllLeft h1.right t1 ) 
            (N.stackAllLeft h2.right t2)
        else if k1  < k2 then -1    
        else 1
    | _, _ -> 0   


let cmp s1 s2 =
  let len1, len2 = N.size s1, N.size s2 in   
  if len1 = len2 then 
    compareAux (N.stackAllLeft s1 []) (N.stackAllLeft s2 [])
  else if len1 < len2 then -1 else 1 

let eq (s1 : t) s2 = 
  cmp s1 s2 = 0





(* This algorithm applies to BST, it does not need to be balanced tree *)  
let rec subset (s1 : t) (s2 : t) =
  match s1, s2 with
    None, _ ->
    true
  | _, None ->
    false
  | Some t1, Some t2 (* Node (l1, v1, r1, _), (Node (l2, v2, r2, _) as t2) *) ->
    let {N.left = l1; value = v1; right = r1 } = t1 in  
    let {N.left = l2; value = v2; right = r2 } = t2 in 
    if v1 = v2 then
      subset l1 l2 && subset r1 r2
    else if v1 < v2 then
      subset (N.create l1 v1 None) l2 && subset r1 s2
    else
      subset (N.create None v1 r1 ) r2 && subset l1 s2


let rec get (n :t) (x : value) = 
  match n with 
  | None -> None
  | Some t  ->    
    let v = t.value in     
    if x = v then Some v
    else get (if x < v then t.left else t.right) x



let rec getUndefined (n :t) (x : value)   = 
  match n with 
  | None -> Js.undefined
  | Some t  ->    
    let v = t.value in     
    if x = v then Js.Undefined.return v
    else getUndefined  (if x < v then t.left else t.right) x

let rec getExn  (n :t) (x : value) = 
  match n with 
  | None -> raise Not_found
  | Some t  ->    
    let v = t.value in     
    if x = v then  v
    else getExn (if x < v then t.left else t.right) x

(****************************************************************************)
let rec addMutate  t  (x : value)=   
  match t with 
  | None -> N.singleton x
  | Some nt -> 
    let k = nt.N.value in 
    if x = k then t 
    else
      let {N.left = l;  right = r} = nt in 
      (if x < k then                   
         nt.left <- addMutate l x       
       else   
         nt.right <- addMutate r x;
      );
      Some (N.balMutate nt)



let fromArray (xs : value array) =   
  let len = A.length xs in 
  if len = 0 then None
  else
    let next =  ref (S.strictlySortedLength xs ) in 
    let result  = 
        ref (
          if next.contents >= 0 then 
            N.fromSortedArrayAux xs 0 next.contents
          else begin 
            next .contents<- - next.contents ;  
            N.fromSortedArrayRevAux xs (next.contents - 1) next.contents
          end
          ) in 
    for i = next.contents to len - 1 do 
      result .contents<- addMutate result.contents (A.getUnsafe xs i) 
    done ;
    result.contents 

File: lib/es6/belt_internalSetInt.js


import * as Belt_SortArrayInt from "./belt_SortArrayInt.js";
import * as Belt_internalAVLset from "./belt_internalAVLset.js";

function has(_t, x) {
  while(true) {
    let t = _t;
    if (t === undefined) {
      return false;
    }
    let v = t.v;
    if (x === v) {
      return true;
    }
    _t = x < v ? t.l : t.r;
    continue;
  };
}

function compareAux(_e1, _e2) {
  while(true) {
    let e2 = _e2;
    let e1 = _e1;
    if (!e1) {
      return 0;
    }
    if (!e2) {
      return 0;
    }
    let h2 = e2.hd;
    let h1 = e1.hd;
    let k1 = h1.v;
    let k2 = h2.v;
    if (k1 !== k2) {
      if (k1 < k2) {
        return -1;
      } else {
        return 1;
      }
    }
    _e2 = Belt_internalAVLset.stackAllLeft(h2.r, e2.tl);
    _e1 = Belt_internalAVLset.stackAllLeft(h1.r, e1.tl);
    continue;
  };
}

function cmp(s1, s2) {
  let len1 = Belt_internalAVLset.size(s1);
  let len2 = Belt_internalAVLset.size(s2);
  if (len1 === len2) {
    return compareAux(Belt_internalAVLset.stackAllLeft(s1, /* [] */0), Belt_internalAVLset.stackAllLeft(s2, /* [] */0));
  } else if (len1 < len2) {
    return -1;
  } else {
    return 1;
  }
}

function eq(s1, s2) {
  return cmp(s1, s2) === 0;
}

function subset(_s1, _s2) {
  while(true) {
    let s2 = _s2;
    let s1 = _s1;
    if (s1 === undefined) {
      return true;
    }
    if (s2 === undefined) {
      return false;
    }
    let v1 = s1.v;
    let l1 = s1.l;
    let r1 = s1.r;
    let v2 = s2.v;
    let l2 = s2.l;
    let r2 = s2.r;
    if (v1 === v2) {
      if (!subset(l1, l2)) {
        return false;
      }
      _s2 = r2;
      _s1 = r1;
      continue;
    }
    if (v1 < v2) {
      if (!subset(Belt_internalAVLset.create(l1, v1, undefined), l2)) {
        return false;
      }
      _s1 = r1;
      continue;
    }
    if (!subset(Belt_internalAVLset.create(undefined, v1, r1), r2)) {
      return false;
    }
    _s1 = l1;
    continue;
  };
}

function get(_n, x) {
  while(true) {
    let n = _n;
    if (n === undefined) {
      return;
    }
    let v = n.v;
    if (x === v) {
      return v;
    }
    _n = x < v ? n.l : n.r;
    continue;
  };
}

function getUndefined(_n, x) {
  while(true) {
    let n = _n;
    if (n === undefined) {
      return;
    }
    let v = n.v;
    if (x === v) {
      return v;
    }
    _n = x < v ? n.l : n.r;
    continue;
  };
}

function getExn(_n, x) {
  while(true) {
    let n = _n;
    if (n !== undefined) {
      let v = n.v;
      if (x === v) {
        return v;
      }
      _n = x < v ? n.l : n.r;
      continue;
    }
    throw new Error("Not_found", {
          cause: {
            RE_EXN_ID: "Not_found"
          }
        });
  };
}

function addMutate(t, x) {
  if (t === undefined) {
    return Belt_internalAVLset.singleton(x);
  }
  let k = t.v;
  if (x === k) {
    return t;
  }
  let l = t.l;
  let r = t.r;
  if (x < k) {
    t.l = addMutate(l, x);
  } else {
    t.r = addMutate(r, x);
  }
  return Belt_internalAVLset.balMutate(t);
}

function fromArray(xs) {
  let len = xs.length;
  if (len === 0) {
    return;
  }
  let next = Belt_SortArrayInt.strictlySortedLength(xs);
  let result;
  if (next >= 0) {
    result = Belt_internalAVLset.fromSortedArrayAux(xs, 0, next);
  } else {
    next = -next | 0;
    result = Belt_internalAVLset.fromSortedArrayRevAux(xs, next - 1 | 0, next);
  }
  for(let i = next; i < len; ++i){
    result = addMutate(result, xs[i]);
  }
  return result;
}

let S;

let N;

let A;

export {
  S,
  N,
  A,
  has,
  compareAux,
  cmp,
  eq,
  subset,
  get,
  getUndefined,
  getExn,
  addMutate,
  fromArray,
}
/* No side effect */
